### DISCUSSION TITLE
How to deploy Medusa using Docker?

### DISCUSSION THREAD

**Discussion by Jengro777** [medusajs/medusa#4125] [Q&A]
How to deploy Medusa using Docker?

**Comment by SebghatYusuf** [comment]
I used this Dockerfile and deployed on GCP Cloud Run and it's working fine, In case it helps: 

```
FROM node:lts

WORKDIR /app/medusa

COPY package.json .
COPY tsconfig.json .

RUN apt-get update

RUN apt-get install -y python

RUN npm install -g npm@latest

RUN yarn global add -g @medusajs/medusa-cli@latest

COPY . .

RUN yarn install

EXPOSE 7000
# ENTRYPOINT ["./develop.sh", "develop"]

# Define the command to run the app
CMD [ "npm", "start" ]
```

**Reply by Jengro777** [reply]
Hello

Thank you very much for your answer



My technical capabilities are limited. Can you provide me with more detailed steps?



Thank you.

**Reply by SebghatYusuf** [reply]
@Avey777 I need to understand where you want to deploy it, because you can have this Dockerfile in root of your project directory and use commands to build the docker image.

**Reply by Jengro777** [reply]
I am deployed on my own server.
Debian11

**Comment by felixelement** [comment]
Hello,

How is your work?

I got this issue:
```
2023-05-29 09:42:25 [ioredis] Unhandled error event: Error: connect ECONNREFUSED 127.0.0.1:6379
2023-05-29 09:42:25     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1487:16)
2023-05-29 09:42:25 Error: connect ECONNREFUSED 127.0.0.1:6379
2023-05-29 09:42:25     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1487:16) {
2023-05-29 09:42:25   errno: -111,
2023-05-29 09:42:25   code: 'ECONNREFUSED',
2023-05-29 09:42:25   syscall: 'connect',
2023-05-29 09:42:25   address: '127.0.0.1',
2023-05-29 09:42:25   port: 6379
2023-05-29 09:42:25 }```

I tried two ways redis server on my local and ubuntu server:

Run by environment
Run in Docker

without docker, it work.

My docker file:

```# docker build -t medusa-docker/node:v1 .
# build stage
FROM node:18.13.0-alpine as build-stage
WORKDIR /app

COPY package*.json package-lock.json ./

RUN npm install @medusajs/admin --legacy-peer-deps
COPY . .
RUN npm -g install cross-env

EXPOSE 9000
CMD ["npm", "start"]
```

My docker compose
```
version: '3.7'

services:
  app:
    image: medusa-docker/node:v1
    restart: unless-stopped
    ports:
      - 9000:9000
    environment:
      JWT_SECRET: ${JWT_SECRET}
      COOKIE_SECRET: ${COOKIE_SECRET}
      MINIO_ENDPOINT: ${MINIO_ENDPOINT}
      MINIO_BUCKET: ${MINIO_BUCKET}
      MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}
      MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}
      MINIO_PRIVATE_BUCKET: ${MINIO_PRIVATE_BUCKET}
      DATABASE_TYPE: ${DATABASE_TYPE}
      DATABASE_URL: ${DATABASE_URL}
      SEGMENT_WRITE_KEY: ${SEGMENT_WRITE_KEY}
      SENDGRID_API_KEY: ${SENDGRID_API_KEY}
      SENDGRID_FROM: ${SENDGRID_FROM}
      SENDGRID_ORDER_PLACED_ID: ${SENDGRID_ORDER_PLACED_ID}
```

Do you have any ideas?

Thank you

**Comment by SGFGOV** [comment]
You are missing the path to redis

On Mon, 29 May, 2023, 19:48 felixelement, ***@***.***> wrote:

> Hello,
>
> How is your work?
>
> I got this issue:
>
> 2023-05-29 09:42:25 [ioredis] Unhandled error event: Error: connect ECONNREFUSED 127.0.0.1:6379
> 2023-05-29 09:42:25     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1487:16)
> 2023-05-29 09:42:25 Error: connect ECONNREFUSED 127.0.0.1:6379
> 2023-05-29 09:42:25     at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1487:16) {
> 2023-05-29 09:42:25   errno: -111,
> 2023-05-29 09:42:25   code: 'ECONNREFUSED',
> 2023-05-29 09:42:25   syscall: 'connect',
> 2023-05-29 09:42:25   address: '127.0.0.1',
> 2023-05-29 09:42:25   port: 6379
> 2023-05-29 09:42:25 }```
>
> I tried two ways redis server on my local and ubuntu server:
>
> Run by environment
> Run in Docker
>
> without docker, it work.
>
> My docker file:
>
>
> docker build -t medusa-docker/node:v1 . build stage
>
> FROM node:18.13.0-alpine as build-stage
> WORKDIR /app
>
> COPY package*.json package-lock.json ./
>
> RUN npm install @medusajs/admin --legacy-peer-deps
> COPY . .
> RUN npm -g install cross-env
>
> EXPOSE 9000
> CMD ["npm", "start"]
>
>
> My docker compose
>
> version: '3.7'
>
> services:
> app:
> image: medusa-docker/node:v1
> restart: unless-stopped
> ports:
> - 9000:9000
> environment:
> JWT_SECRET: ${JWT_SECRET}
> COOKIE_SECRET: ${COOKIE_SECRET}
> MINIO_ENDPOINT: ${MINIO_ENDPOINT}
> MINIO_BUCKET: ${MINIO_BUCKET}
> MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}
> MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}
> MINIO_PRIVATE_BUCKET: ${MINIO_PRIVATE_BUCKET}
> DATABASE_TYPE: ${DATABASE_TYPE}
> DATABASE_URL: ${DATABASE_URL}
> SEGMENT_WRITE_KEY: ${SEGMENT_WRITE_KEY}
> SENDGRID_API_KEY: ${SENDGRID_API_KEY}
> SENDGRID_FROM: ${SENDGRID_FROM}
> SENDGRID_ORDER_PLACED_ID: ${SENDGRID_ORDER_PLACED_ID}
>
>
> Do you have any ideas?
>
> Thank you
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/medusajs/medusa/discussions/4125#discussioncomment-6029765>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXEQJHDVFGR3QFKVBMJEUNTXISV2FANCNFSM6AAAAAAYGJL4JE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

-- 

This message (including any attachments) may contain confidential, 
proprietary, privileged and/or private
information. The information is 
intended to be for the use of the individual or entity designated above. If
you are not the intended recipient of this message, please notify the 
sender immediately, and delete the
message and any attachments. Any 
disclosure, reproduction, distribution or other use of this message or
any 
attachments by an individual or entity other than the intended recipient is 
prohibited.

**Comment by felixelement** [comment]
Hi @SGFGOV thanks for your answer,

i added

```version: '3.7'

services:
  app:
    image: medusa-docker/node:v1
    restart: unless-stopped
    ports:
      - 9000:9000
    environment:
      JWT_SECRET: ${JWT_SECRET}
      COOKIE_SECRET: ${COOKIE_SECRET}
      MINIO_ENDPOINT: ${MINIO_ENDPOINT}
      MINIO_BUCKET: ${MINIO_BUCKET}
      MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}
      MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}
      MINIO_PRIVATE_BUCKET: ${MINIO_PRIVATE_BUCKET}
      DATABASE_TYPE: ${DATABASE_TYPE}
      DATABASE_URL: ${DATABASE_URL}
      SEGMENT_WRITE_KEY: ${SEGMENT_WRITE_KEY}
      SENDGRID_API_KEY: ${SENDGRID_API_KEY}
      SENDGRID_FROM: ${SENDGRID_FROM}
      SENDGRID_ORDER_PLACED_ID: ${SENDGRID_ORDER_PLACED_ID}
      REDIS_URL: ${REDIS_URL}
```.

However, it did not work too.

Could you please tell me more details?

I do appreciate it.

Thank you so much

**Comment by SGFGOV** [comment]
Can you share thr redis Url?

On Mon, 29 May, 2023, 20:11 felixelement, ***@***.***> wrote:

> Hi @SGFGOV <https://github.com/SGFGOV> thanks for your answer,
>
> i added
>
>
> services:
>   app:
>     image: medusa-docker/node:v1
>     restart: unless-stopped
>     ports:
>       - 9000:9000
>     environment:
>       JWT_SECRET: ${JWT_SECRET}
>       COOKIE_SECRET: ${COOKIE_SECRET}
>       MINIO_ENDPOINT: ${MINIO_ENDPOINT}
>       MINIO_BUCKET: ${MINIO_BUCKET}
>       MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}
>       MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}
>       MINIO_PRIVATE_BUCKET: ${MINIO_PRIVATE_BUCKET}
>       DATABASE_TYPE: ${DATABASE_TYPE}
>       DATABASE_URL: ${DATABASE_URL}
>       SEGMENT_WRITE_KEY: ${SEGMENT_WRITE_KEY}
>       SENDGRID_API_KEY: ${SENDGRID_API_KEY}
>       SENDGRID_FROM: ${SENDGRID_FROM}
>       SENDGRID_ORDER_PLACED_ID: ${SENDGRID_ORDER_PLACED_ID}
>       REDIS_URL: ${REDIS_URL}
> ```.
>
> However, it did not work too.
>
> Could you please tell me more details?
>
> I do appreciate it.
>
> Thank you so much
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/medusajs/medusa/discussions/4125#discussioncomment-6029962>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXEQJHAVTOU3JXMFE56OR7TXISYTFANCNFSM6AAAAAAYGJL4JE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 

This message (including any attachments) may contain confidential, 
proprietary, privileged and/or private
information. The information is 
intended to be for the use of the individual or entity designated above. If
you are not the intended recipient of this message, please notify the 
sender immediately, and delete the
message and any attachments. Any 
disclosure, reproduction, distribution or other use of this message or
any 
attachments by an individual or entity other than the intended recipient is 
prohibited.

**Reply by SGFGOV** [reply]
Typically it is redis://default:password@redisHost:redisport

Sometimes you can omit default

**Reply by felixelement** [reply]
REDIS_URL=redis://localhost:6379 here it is. Thanks

**Comment by felixelement** [comment]
Hello I found the way to make it work. I used my ip:6379 instead of localhost:6379

Hope it help

**Reply by Jengro777** [reply]
Could you kindly provide the complete content of the docker-compose.yml file for me to review?

**Reply by felixelement** [reply]
# docker-compose --env-file .env up -d
version: '3.7'

services:
  app:
    image: medusa-docker/node:v1
    restart: unless-stopped
    ports:
      - 9000:9000
    environment:
      JWT_SECRET: ${JWT_SECRET}
      COOKIE_SECRET: ${COOKIE_SECRET}
      MINIO_ENDPOINT: ${MINIO_ENDPOINT}
      MINIO_BUCKET: ${MINIO_BUCKET}
      MINIO_ACCESS_KEY: ${MINIO_ACCESS_KEY}
      MINIO_SECRET_KEY: ${MINIO_SECRET_KEY}
      MINIO_PRIVATE_BUCKET: ${MINIO_PRIVATE_BUCKET}
      DATABASE_TYPE: ${DATABASE_TYPE}
      DATABASE_URL: ${DATABASE_URL}
      SEGMENT_WRITE_KEY: ${SEGMENT_WRITE_KEY}
      SENDGRID_API_KEY: ${SENDGRID_API_KEY}
      SENDGRID_FROM: ${SENDGRID_FROM}
      SENDGRID_ORDER_PLACED_ID: ${SENDGRID_ORDER_PLACED_ID}
      REDIS_URL: ${REDIS_URL}

**Reply by Jengro777** [reply]
Can you share the configuration content of .env? I don't know how to fill in the information inside.

**Comment by Jengro777** [comment]
Is this it?
https://hub.docker.com/r/misasa/medusa/tags

**Comment by Jengro777** [comment]
This is too difficult for me. I'm thinking of giving up on using Medusa.

**Reply by Jengro777** [reply]
Why not upload the Docker images to Hub Docker?

**Comment by predaytor** [comment]
I deployed to fly.io (Fly Postgres, Fly Upstash Redis, Fly Volumes for uploads storage). I use Docker only in production, for dev all you need is `yarn dev`.

`Dockerfile`:
```Dockerfile
FROM node:22-alpine AS base
LABEL fly_launch_runtime="Node.js"
WORKDIR /usr/src/app

# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat


# Install dependencies only when needed
FROM base AS deps

# Install dependencies based on the preferred package manager
COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./

RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then corepack enable pnpm && pnpm i --frozen-lockfile; \
  else echo "Lockfile not found." && exit 1; \
  fi


# Rebuild the source code only when needed
FROM base AS builder

ENV NODE_ENV production

COPY --from=deps /usr/src/app/node_modules ./node_modules
COPY . .

RUN \
  if [ -f yarn.lock ]; then yarn run build; \
  elif [ -f package-lock.json ]; then npm run build; \
  elif [ -f pnpm-lock.yaml ]; then corepack enable pnpm && pnpm run build; \
  else echo "Lockfile not found." && exit 1; \
  fi


# Final stage for app image
FROM base as runner

# Setting NODE_ENV=production requires secure cookie (https only access)
ENV NODE_ENV production
ENV PORT 9000
ENV HOST 0.0.0.0

# Copy config file
COPY ./medusa-config.js ./

# Copy built application
COPY --from=builder /usr/src/app/package.json ./package.json
COPY --from=builder /usr/src/app/node_modules ./node_modules
COPY --from=builder /usr/src/app/build ./build
COPY --from=builder /usr/src/app/dist ./dist

# Run application
EXPOSE 9000
CMD ["npm", "run", "start"]
```
`package.json` (change default start script):
```json
{
    "scripts": {
        "start": "medusa start --host=0.0.0.0 --port=9000",
        "migrate": "medusa migrations run"
    }
}
```
`medusa-config.js` (I do not use `autoRebuild`):
```js
const dotenv = require('dotenv');

let ENV_FILE_NAME = '';

switch (process.env.NODE_ENV) {
	case 'production':
		ENV_FILE_NAME = '.env';
		break;
	case 'staging':
		ENV_FILE_NAME = '.env.staging';
		break;
	default:
		ENV_FILE_NAME = '.env.local';
		break;
}

try {
	dotenv.config({ path: process.cwd() + '/' + ENV_FILE_NAME });
} catch (e) {}

// CORS when consuming Medusa from admin
const ADMIN_CORS = process.env.ADMIN_CORS || 'http://localhost:7001';

// CORS to avoid issues when consuming Medusa from a client
const STORE_CORS = process.env.STORE_CORS || 'http://localhost:3000';

const DATABASE_URL = process.env.DATABASE_URL || 'postgres://localhost/medusa-starter-default';

const REDIS_URL = process.env.REDIS_URL || 'redis://localhost:6379';

const plugins = [
	`medusa-fulfillment-manual`,
	`medusa-payment-manual`,
	{
		resolve: '@medusajs/admin',
		/** @type {import('@medusajs/admin').PluginOptions} */
		options: {
			// autoRebuild: true,
			path: '/app',
			develop: {
				open: false,
				allowedHosts: 'all',
			},
		},
	},
	{
		resolve: `@medusajs/file-local`,
		options: {
			upload_dir: 'uploads',
			backend_url: process.env.MEDUSA_BACKEND_URL || 'http://localhost:9000',
		},
	},
	{
		resolve: `medusa-payment-stripe`,
		options: {
			api_key: process.env.STRIPE_API_KEY,
			webhook_secret: process.env.STRIPE_WEBHOOK_SECRET,
		},
	},
];

const modules = {
	eventBus: {
	    resolve: "@medusajs/event-bus-redis",
	    options: {
	         redisUrl: REDIS_URL
	    }
	},
	cacheService:
		process.env.NODE_ENV !== 'development'
			? {
					resolve: '@medusajs/cache-redis',
					options: {
						redisUrl: REDIS_URL,
						ttl: 30,
					},
				}
			: {},
};

/** @type {import('@medusajs/medusa').ConfigModule["projectConfig"]} */
const projectConfig = {
	jwtSecret: process.env.JWT_SECRET,
	cookieSecret: process.env.COOKIE_SECRET,
	store_cors: STORE_CORS,
	admin_cors: ADMIN_CORS,
	database_url: DATABASE_URL,
	redis_url: process.env.NODE_ENV !== 'development' ? REDIS_URL : undefined,
	database_extra:
		process.env.NODE_ENV !== 'development'
			? {
					ssl: {
						rejectUnauthorized: false,
					},
				}
			: {},
};

/** @type {import('@medusajs/medusa').ConfigModule} */
module.exports = {
	projectConfig,
	plugins,
	modules,
};
```
`fly.toml`:
```toml
app = 'my-app-store'
primary_region = 'lhr'

[build]

[deploy]
  release_command = 'npm run migrate'

[[mounts]]
  source = 'my_app_uploads'
  destination = '/usr/src/app/uploads'

[http_service]
  internal_port = 9000
  force_https = true
  auto_stop_machines = 'suspend'
  auto_start_machines = true
  min_machines_running = 1
  processes = ['app']

[[vm]]
  memory = '2gb'
  cpu_kind = 'shared'
  cpus = 1
```
`.env`:
```env
DATABASE_TYPE=postgres
DATABASE_URL=postgres://****:5432/smme?sslmode=disable

REDIS_URL=redis://****:6379?family=6

MEDUSA_BACKEND_URL=https://example.com
MEDUSA_ADMIN_ONBOARDING_TYPE=default

# ADMIN_CORS='http://localhost:9000,http://localhost:7001'
# STORE_CORS='http://localhost:8000,http://localhost:7001'

JWT_SECRET=secret
COOKIE_SECRET=secret

STRIPE_API_KEY=sk_****
# only necessary for production
STRIPE_WEBHOOK_SECRET=****
```
